feat: add actor list

niermann999 opened the pull request:
This PR adds an actor list that allows call dependencies between actors. These are resolved at compile time by a composition pattern, where the components of a composite are the observers of the composite's own primary actor implementation. It is however impossible to observe observers of another composition, limiting the call depth to a single level.

The states of the actors are ordered in a flat tuple (not according to call order a 'tuple of tuple of ...' ), that has to be built outside of the ```actor_chain```. Like this, the states can be transferred to device individually and the tuple, as well as the the actors and the ```actor_chain```, can be assembled on device.  With this it is also possible to have multiple actors share a single state, e.g. if it is read-only or empty. The actor states are, however, not owned by the ```propagator::state```, but are passed to it as a tuple of references.

Also taught ```thrust::get``` to resolve a type instead of an index. Together with tagging class types (e.g. ```is_comp_actor```) this has reduced the code complexity quite a bit (even though as beomki observed, the actor chain also works with ```std::tuple``` on device). Actors now need to inherit from the ```detray::actor```, define a ```state_type``` and be callable on their state type, the ```propagator::state```, as well as optionally the state of their subject actor. The propagation inspectors were now turned into actors as a first test and to shorten the ```propagate``` interface a bit.

I compiled and tested it on the EP threadripper machine now.

beomki-yeo commented:
TBH, I am afraid that this PR over-complicates the user interface.

Could you explain the following:
1. Motivation of this PR. and how is it better than the original inspector?
2. Why do we need `actor_state`?
3. Why do we need `composite_actor` when we can compose actors in `actor_chain`?

niermann999 commented:
>     1. Motivation of this PR. and how is it better than the original inspector?

This PR is of course not meant to refactor the inspector classes, but to provide a general infrastructure for any kind of class that needs to be called during propagation. I don't think we want to pass them all as function parameters. Since the inspectors fit the same usecase, I implemented them as actors as an example and a test.

>     2. Why do we need `actor_state`?

In case an actor needs to hold data per track or needs to pass on results for a track. It is modeled after the acts ```result``` type. I followed the design of the stepper and the navigator here, to completely separate the state from the class, because I think it is a neat design.

>     3. Why do we need `composite_actor` when we can compose actors in `actor_chain`?

I do agree that unfortunately the implementation of the ```composite_actor``` turned out way more complicated than I had expected, but given that it just implements two of the well known patterns (composition and observer) I think its still readable. I implemented it, to be able to have actors be informed about the results of another actor, without having to implement this logic inside the actors themselves. They should not know how to obtain their and their subjects state. I would like to experiment with chained actors a bit in the navigation policies. It is also possible with this design to call special aborters for actors (e.g. the KalmanAborter, that needs the updated state of the Kalman filter).

While I agree that the implementation of the ```actor_chain``` has turned out a bit complicated, I am somewhat surprised that you think the user interface is complicated. I think the changes in the ```propagator``` are minimal and I find the requirement that an actor needs to be callable on its state and the propagation state to be reasonable?
A lot of the complication in the interface actually stems from the fact, that I have to resolve an actor ```ID``` to get its state out of the state (thrust) tuple. If we could get a layer of abstraction in here, some of the ```run``` function overloads would falll away, as well as the necessity to order the tuple etc.

niermann999 commented:
Ok, no, nvm. With templated actor state types, fetching them by ```ID``` is the only possibility I can come up with.

beomki-yeo commented on tests/common/include/tests/common/tools_propagator.inl:
```diff
@@ -131,6 +238,10 @@ TEST_P(PropagatorWithRkStepper, propagator_rk_stepper) {
     // Set origin position of tracks
     const point3 ori{0., 0., 0.};
 
+    auto actor_states =
```
Could you define `actor_chain_t::state` in actor_chain struct so you don't have to write down aggregation twice?

niermann999 replied:
I am not sure I understand. I still have to call the constructors of the states, right?

beomki-yeo replied:
Right

beomki-yeo commented on core/include/detray/propagator/propagator.hpp:
```diff
@@ -56,15 +45,20 @@ struct propagator {
         : _stepper(std::move(s)), _navigator(std::move(n)) {}
 
     /** Propagation that state aggregates a stepping and a navigation state */
+    template <typename actor_states_t = bool>
```
Can we remove this `actor_state_t` from template parameter and use `actor_chaint_t::state` instead?

niermann999 replied:
Good idea!

beomki-yeo commented on core/include/detray/propagator/actor_chain.hpp:
```diff
@@ -0,0 +1,96 @@
+    public:
+    /// Types of the actors that are registered in the chain
+    using actor_list_type = tuple_t<actors_t...>;
+
```
Can we define `actor_state` here?
`using actor_state = tuple_t<actors_t::state ...>`

niermann999 replied:
Done

niermann999 commented:
Hm, it seems switching to a state tuple of references has slowed the code quite a bit...

beomki-yeo commented on tests/common/include/tests/common/tools/helix_gun.hpp:
```diff
@@ -17,11 +17,54 @@ class helix_gun {
+    /// Sets up a new helix from
+    /// @param vertex the particle origin and
+    /// @param mag_field a b field access.
+    void init(const free_track_parameters vertex,
```
Do we really need this function?

beomki-yeo commented on tests/common/include/tests/common/tools_propagator.inl:
```diff
@@ -150,18 +249,21 @@ TEST_P(PropagatorWithRkStepper, propagator_rk_stepper) {
-            combined_inspector ci{helix_inspector(helix),
-                                  propagation::print_inspector{}};
+            std::get<helix_inspector::state_type &>(actor_states)
+                ._helix.init(traj, &B);
```
I think you have to use the original constructor for `helix_gun`. Or is there any technical issue with that

niermann999 replied:
See comment :) I just could not get it to work back then. Let's have another try

niermann999 replied:
Ok, it works now. I guess, using a reference to the b field might have also deleted the copy assignment operator previously. Thanks for pointing out!

beomki-yeo reviewed (commented):
Looks good in general. Could you explain why you are defining default values for some struct such as `track` or `helix_gun`?

beomki-yeo commented on core/include/detray/definitions/detail/accessor.hpp:
```diff
@@ -35,6 +36,26 @@ namespace vtuple = std;
 namespace detail {
 
+namespace {
+struct empty_type {};
```
This doesn't seem to be used

beomki-yeo replied:
Ah it is used below..

beomki-yeo replied:
NVM it is used

beomki-yeo commented on tests/unit_tests/cuda/propagator_cuda_kernel.hpp:
```diff
@@ -58,33 +56,53 @@ constexpr scalar pos_diff_tolerance = 1e-3;
+using inspector_host_t = track_inspector<vecmem::vector>;
+using inspector_device_t = track_inspector<vecmem::device_vector>;
+using actor_chain_host_t = actor_chain<thrust::tuple, inspector_host_t>;
+using actor_chain_device_t = actor_chain<thrust::tuple, inspector_host_t>;
```
Have you tried using `std::tuple` for this?

niermann999 replied:
Actually no. i was under the impression that ```thrust::tuple``` was also used host-side in the cuda tests. I can try ```std::tuple```, but that is basically the ```common/test/propagator.inl``` test, right? I think, it might be time to define proper headers in ```common/test/*```. What do you think?

niermann999 replied:
Nice, seems it actually runs with ```std::tuple```. Although at this point it does not make much of a difference for this PR anymore (except compilation time I guess). Regardless, I removed the template now :)

beomki-yeo replied:
Ah OK. It's good to confirm it is working with `std::tuple` 
Actually I think it is better to keep the tuple template for the moment. (as `std::tuple` might not work well without `--expt-relaxed-constexpr`)
Could you roll back one commit?

niermann999 replied:
Sure

niermann999 commented:
> Looks good in general. Could you explain why you are defining default values for some struct such as `track` or `helix_gun`? Is it related to `thrust::tuple` issue you mentioned above?

Cool! Hm, no, the problem was that the default constructors are implicitly deleted when a parametrized constructor is provided, so I tried adding it by hand. I did this when I tried to default construct the ```helix_inspector```, but none of it really worked well, so I settled for the ```init()``` method there. Let's remove the default values again.

niermann999 commented on tests/common/include/tests/common/tools_propagator.inl:
```diff
@@ -24,6 +26,130 @@
+}  // anonymous namespace
+
+// Test the actor chain on some dummy actor types
+TEST(ALGEBRA_PLUGIN, actor_chain) {
```
Good idea

beomki-yeo reviewed (approved):
Let's get it merged once you roll back one commit...

niermann999 commented:
Correction, the bug in the CUDA array rk-stepping test seems to be introduced here. Let me have a look

niermann999 commented:
What happens with the CI remnants now? I cannot merge this, can you @asalzburger ?

beomki-yeo commented:
I thought it would be cleaned up after merging #233 but it didn't ...

beomki-yeo commented:
https://github.community/t/github-actions-pull-requests-always-include-some-outdated-checks/16157

Looks like the maintainer who has an access to **Setting** category (so maybe @asalzburger or @paulgessinger) should disable the outdated workflows

paulgessinger commented:
Ok I manually went through and manually clicked all of them.

Can we trim the build matrix though? Do we really need this entire matrix?

beomki-yeo commented:
I agree that there are too many workflows. I can remove double build type in container
